src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,417 @@
+        self.get_box(input_box)
+            .borrow_mut()
+            .ranging_quantifiers
+            .insert(id);
+        id
+    }
+
+    fn get_quantifier(&self, quantifier_id: BoxId) -> &RefCell<Quantifier> {
+        &*self.quantifiers.get(&quantifier_id).unwrap()
+    }
+
+    /// Visit boxes in the query graph in pre-order
+    fn visit_pre_boxes<F, E>(&self, f: &mut F) -> Result<(), E>
+    where
+        F: FnMut(&RefCell<QueryBox>) -> Result<(), E>,
+    {
+        self.visit_pre_boxes_in_subgraph(f, self.top_box)
+    }
+
+    /// Visit boxes in the query graph in pre-order
+    fn visit_pre_boxes_in_subgraph<F, E>(&self, f: &mut F, start_box: BoxId) -> Result<(), E>
+    where
+        F: FnMut(&RefCell<QueryBox>) -> Result<(), E>,
+    {
+        let mut visited = HashSet::new();
+        let mut stack = vec![start_box];
+        while !stack.is_empty() {
+            let box_id = stack.pop().unwrap();
+            if visited.insert(box_id) {
+                let query_box = self.get_box(box_id);
+                f(query_box)?;
+
+                stack.extend(
+                    query_box
+                        .borrow()
+                        .quantifiers
+                        .iter()
+                        .rev()
+                        .map(|q| self.get_quantifier(*q).borrow().input_box),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Removes unreferenced objects from the model. May be invoked
+    /// several times during query rewrites.
+    #[allow(dead_code)]
+    fn gargabe_collect(&mut self) {
+        let mut visited_boxes = HashSet::new();
+        let mut visited_quantifiers = HashSet::new();
+        let mut stack = vec![self.top_box];
+        while !stack.is_empty() {
+            let box_id = stack.pop().unwrap();
+            if visited_boxes.insert(box_id) {
+                let query_box = self.get_box(box_id);
+                for q in query_box.borrow().quantifiers.iter() {
+                    visited_quantifiers.insert(*q);
+                    let q = self.get_quantifier(*q).borrow();
+                    stack.push(q.input_box);
+                }
+            }
+        }
+        self.boxes.retain(|b, _| visited_boxes.contains(b));
+        self.quantifiers
+            .retain(|q, _| visited_quantifiers.contains(q));
+    }
+}
+
+/// A semantic operator within a Query Graph.
+#[derive(Debug)]
+pub struct QueryBox {
+    /// uniquely identifies the box within the model
+    pub id: BoxId,
+    /// the type of the box
+    pub box_type: BoxType,
+    /// the projection of the box
+    pub columns: Vec<Column>,
+    /// the input quantifiers of the box
+    pub quantifiers: QuantifierSet,
+    /// quantifiers ranging over this box
+    pub ranging_quantifiers: QuantifierSet,
+    /// whether this box must enforce the uniqueness of its output, it is
+    /// guaranteed by structure of the box or it must preserve duplicated
+    /// rows from its input boxes.
```

Comment:
Can you label `DistinctOperation` with what each variant means and then have this comment be "See [DistinctOperation]"?